pkg/operator/oidcdiscoveryendpoint/controller.go

```diff
@@ -0,0 +1,424 @@
+	eventRecorder      events.Recorder
+	infrastructureName string
+	region             string
+}
+
+var _ reconcile.Reconciler = &s3EndpointReconciler{}
+
+func (r *s3EndpointReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	r.logger.Info("reconciling AWS S3 OIDC discovery endpoint")
+
+	isUnmanaged, err := r.reconcileClusterAuthentication()
+	if err != nil {
+		r.logger.WithError(err).Error("failed reconciling cluster authentication")
+		return reconcile.Result{Requeue: true}, err
+	}
+	if isUnmanaged {
+		return reconcile.Result{}, nil
+	}
+
+	err = r.reconcileS3Resources()
+	if err != nil {
+		r.logger.WithError(err).Error("failed reconciling S3 resources")
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *s3EndpointReconciler) reconcileClusterAuthentication() (bool, error) {
+	r.logger.Debugf("reconciling cluster Authentication ServiceAccountIssuer")
+
+	auth, err := r.configclientset.ConfigV1().Authentications().Get(context.TODO(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	serviceAccountIssuer := r.getIssuerURL()
+	if auth.Spec.ServiceAccountIssuer != "" {
+		r.logger.Debug("Authentication ServiceAccountIssuer already specified")
+		if auth.Spec.ServiceAccountIssuer != serviceAccountIssuer {
+			r.logger.Info("Authentication ServiceAccountIssuer is not set to the s3 bucket location, skipping OIDC reconciliation")
+			return true, nil
+		}
+		return false, nil
+	}
+
+	auth.Spec.ServiceAccountIssuer = serviceAccountIssuer
+	_, err = r.configclientset.ConfigV1().Authentications().Update(context.TODO(), auth, metav1.UpdateOptions{})
+	return false, err
+}
+
+func (r *s3EndpointReconciler) getIssuerURL() string {
+	return fmt.Sprintf("https://s3-%s.amazonaws.com/%s", r.region, r.getBucketName())
+}
+
+func (r *s3EndpointReconciler) getBucketName() string {
+	return fmt.Sprintf("%s-oidc", r.infrastructureName)
+}
+
+func (r *s3EndpointReconciler) reconcileS3Resources() error {
+	r.logger.Debugf("reconciling S3 resources")
+
+	// Get the root secret and create an AWS client
+	secret, err := r.kubeclientset.CoreV1().Secrets(credentialsSecretNamespace).Get(context.TODO(), credentialsSecretName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	accessKey, ok := secret.Data[awsannotator.AwsAccessKeyName]
+	if !ok {
+		return fmt.Errorf("couldn't fetch key containing %s from cloud cred secret", awsannotator.AwsAccessKeyName)
+	}
+
+	secretKey, ok := secret.Data[awsannotator.AwsSecretAccessKeyName]
+	if !ok {
+		return fmt.Errorf("couldn't fetch key containing %s from cloud cred secret", awsannotator.AwsSecretAccessKeyName)
+	}
+
+	awsClient, err := aws.NewClient(accessKey, secretKey, r.region, r.infrastructureName)
+	if err != nil {
+		return fmt.Errorf("error creating aws client: %v", err)
+	}
+
+	// Ensure bucket exists
+	bucketName := r.getBucketName()
+	_, err = awsClient.CreateBucket(&s3.CreateBucketInput{
```

Comment:
have we though about how GET->CREATE vs CREATE only affects our rate limiting? Maybe AWS allows a lot more GETs and that would prevent us from sending CREATE, while depending on CREATE to check if exists might trigger more rate limiting.

Reply:
the reconciliation happens very infrequently.  the only thing that can trigger it is a change to the service account token signer (which @marun informs me doesn't rotate on its own at any interval).  even if we were to get in a failure loop, the retry has backoff so we shouldn't cause any issues wrt rate limiting / throttling

Reply:
if you can add a note to the commit message on this, it will help in easy discoverability.  github comments aren't great after the fact.